README.md
The Bugsnag agent forwards reports from your server to bugsnag.com or an
on-premise Bugsnag installation. It's used to avoid any latency impact that
might occur if you need to make a call over the network in every exception
handler.

## Getting started

First you need to install the `bugsnag-agent` binary, and make it executable on your server:

```bash
$ pip install bugsnag-agent
```

Next you need to run it.

```bash
$ bugsnag-agent
Bugsnag Agent started. http://127.0.0.1:3829 -> https://notify.bugsnag.com/
```

You can verify that it's running using curl in another terminal:

```bash
$ curl http://127.0.0.1:3829
Bugsnag agent: 127.0.0.1:3829 -> https://notify.bugsnag.com/ (0/1000)%
```

Finally you need to configure the endpoint of your Bugsnag apps to be `http://localhost:3829`. This differs per [notifier](https://docs.bugsnag.com/platforms/), but for example PHP is:

```php
$bugsnag->setEndpoint("localhost:3829");
$bugsnag->setUseSSL(false);
```

## Upstart

If you'd like to ensure Bugsnag is always running, you can save the following `upstart` script as `/etc/init/bugsnag.conf` and then run `sudo start bugsnag`.

```upstart
#!upstart
description "Bugsnag forwarding agent"
author      "Bugsnag <support@bugsnag.com>"

start on (filesystem and net-device-up IFACE=lo)
stop on shutdown

respawn
respawn limit 99 5

console log

script
    exec bugsnag-agent
end script
```

## Systemd

If you'd like to configure Bugsnag to be managed by `systemd` instead, drop the following into `/etc/systemd/system/bugsnag-agent.service`:

```
[Unit]
Description="Bugsnag Agent"
After=network.target

[Service]
Type=simple
ExecStart=/usr/local/bin/bugsnag-agent

[Install]
WantedBy=multi-user.target
```

After that, reload the `systemd` daemon, enable the service (so that it runs at boot) and start it:

```bash
sudo systemctl daemon-reload
sudo systemctl enable bugsnag-agent.service
sudo systemctl start bugsnag-agent.service
```

## Configuration

`bugsnag-agent` reads /etc/bugsnag.conf if it exists. The default configuration is:

```conf
# Default configuration for bugsnag.
[bugsnag]

# The port to listen on.
port = 3829

# The interfact to listen on. Set this to 0.0.0.0 if you want to allow anyone
# to forward to Bugsnag. If you do that, ensure that this process is firewalled
# off from the global internet.
listen = 127.0.0.1

# The endpoint to send exceptions to. This can be another `bugsnag-agent`,
# https://notify.bugsnag.com/ or your local on-premise Bugsnag.
endpoint = https://notify.bugsnag.com/

# Logging verbosity
log_level = INFO
```

You can change which configuration file is used with the `-c` parameter:

```
bugsnag-agent -c /usr/local/etc/bugsnag.conf
```

All the options can also be set directly using command line flags:

```
bugsnag-agent --port 3829 --listen 127.0.0.1 --endpoint https://notify.bugsnag.com/ --log-level INFO
```

## Meta
`bugsnag-agent` is copyright under the MIT license. See LICENSE.MIT for details.

Bug reports and pull requests are welcome.


CHANGELOG.md
## 1.3.0

### Enhancements

* Added support for Python 3
  [#27](https://github.com/bugsnag/bugsnag-agent/pull/27)
  [Birgir Haraldsson](https://github.com/bix0r)

### Bug fixes

* Include HTTP headers when retrying requests if an error is raised
  [#27](https://github.com/bugsnag/bugsnag-agent/pull/27)
  [Birgir Haraldsson](https://github.com/bix0r)

## 1.2.1

### Enhancements

* Added support for CORS for agents running in different host
  [#20](https://github.com/bugsnag/bugsnag-agent/pull/20)

## 1.2.0

### Enhancements

* Added support for required Bugsnag header forwarding
  [#18](https://github.com/bugsnag/bugsnag-agent/pull/18)

## 1.1.0

Package for PyPI

### Bug fixes

* Improve error handling for malformed requests
  [#14](https://github.com/bugsnag/bugsnag-agent/pull/14)
  [Birgir Haraldsson](https://github.com/bix0r)

## 1.0.0

Initial release


agent.py
from optparse import OptionParser
from multiprocessing import Process
from ConfigParser import RawConfigParser
from Queue import Queue
from urlparse import urlparse
from time import sleep
from BaseHTTPServer import BaseHTTPRequestHandler, HTTPServer
from threading import Thread
from thread import interrupt_main
from traceback import print_exception
import httplib
import sys

class BugsnagAgent(object):
    """
        The BugsnagAgent sits on your server and forwards exception payloads to
        https://notify.bugsnag.com/.

        It's designed to protect you against any latency spikes that may
        occur talking accross the internet in an exception handler.
    """

    def __init__(self):
        self.parse_config()
        self.queue = Queue(1000)

    def parse_config(self):
        """
            initializes self.port, self.listen, self.endpoint and self.connection
        """

        default = {
            "endpoint": "https://notify.bugsnag.com/",
            "port": 3829,
            "listen": "127.0.0.1",
        }

        parser = OptionParser()
        parser.add_option(
            "-c", "--config",
            dest="config_file",
            default="/etc/bugsnag.conf",
            help="The path to your config file (default /etc/bugsnag.conf)"
        )
        parser.add_option(
            "-e", "--endpoint",
            dest="endpoint",
            help=("the URL of your Bugsnag server (default %s)" % default['endpoint'])
        )
        parser.add_option(
            "-p", "--port",
            dest="port",
            help=("the port to bind to (default %s)" % default['port'])
        )
        parser.add_option(
            "-l", "--listen",
            dest="listen",
            help=("the ip to listen to (use 0.0.0.0 to allow anyone to connect, default %s)" % default["listen"])
        )

        (options, args) = parser.parse_args()

        config = RawConfigParser()
        config.read(options.config_file)

        if options.endpoint is not None:
            self.endpoint = options.endpoint
        elif config.has_option("bugsnag", "endpoint"):
            self.endpoint = config.get("bugsnag", "endpoint")
        else:
            self.endpoint = default["endpoint"]

        endpoint = urlparse(self.endpoint)
        if endpoint.scheme == 'https':
            self.connection = httplib.HTTPSConnection(endpoint.hostname, endpoint.port or 443)
        else:
            self.connection = httplib.HTTPConnection(endpoint.hostname, endpoint.port or 80)

        if options.port is not None:
            self.port = options.port
        elif config.has_option("bugsnag", "port"):
            self.port = config.get("bugsnag", "port")
        else:
            self.port = default["port"]

        if options.listen is not None:
            self.listen = options.listen
        elif config.has_option("bugsnag", "listen"):
            self.listen = config.get("bugsnag", "listen")
        else:
            self.listen = default['listen']

    def start(self):
        """
            Run the agent, and wait for a SIGINT or SIGTERM.
        """
        try:
            server = Thread(target=self._thread(self._server), name="server")
            client = Thread(target=self._thread(self._client), name="client")
            server.setDaemon(True)
            client.setDaemon(True)
            server.start()
            client.start()

            print "Bugsnag Agent started. http://%s:%s -> %s" % (self.listen, self.port, self.endpoint)
            while True:
                sleep(1000)
        except KeyboardInterrupt:
            # give threads time to print exceptions
            sleep(0.1)

    def enqueue(self, body):
        """
            Add a new payload to the queue.
        """
        try:
            self.queue.put_nowait(body)
            print "Enqueued %s bytes (%s/%s)" % (len(body), self.queue.qsize(), self.queue.maxsize)
        except:
            print "Discarding report as queue is full: %s" % (repr(body))

    def _server(self):
        """
            Run the HTTP server on (self.listen, self.port) that puts
            payloads into the queue.
        """
        server = HTTPServer((self.listen, self.port), BugsnagHTTPRequestHandler)
        server.bugsnag = self
        server.serve_forever()

    def _client(self):
        """
            Continually monitor the queue and send anything in it to Bugsnag.
        """
        while True:
            body = self.queue.get(True)
            print "Sending %s bytes (%s/%s)" % (len(body), self.queue.qsize(), self.queue.maxsize)

            try:
                self.connection.request("POST", "/", body)
                self.connection.getresponse()
                self.connection.close()
            except:
                print "Cannot send request. Retrying in 5 seconds"
                print_exception(*sys.exc_info())
                print "continuing..."
                self.connection.close()
                self.enqueue(body)
                sleep(5)

    def _thread(self, target):
        def run():
            try:
                target()
            except:
                interrupt_main()
                print_exception(*sys.exc_info())
            pass
        return run

class BugsnagHTTPRequestHandler(BaseHTTPRequestHandler):

    def do_GET(self):
        """
            Show the current status of the agent
        """
        self.send_response(200)
        self.end_headers()
        bugsnag = self.server.bugsnag
        self.wfile.write("Bugsnag agent: %s:%s -> %s (%s/%s)" % (bugsnag.listen, bugsnag.port, bugsnag.endpoint, bugsnag.queue.qsize(), bugsnag.queue.maxsize))

    def do_POST(self):
        """
            Accept a payload for forwarding
        """
        bugsnag = self.server.bugsnag
        body = self.rfile.read(int(self.headers['Content-Length']))
        bugsnag.enqueue(body)

        response = ("OK %s:%s -> %s (%s/%s)\n" % (bugsnag.listen, bugsnag.port, bugsnag.endpoint, bugsnag.queue.qsize(), bugsnag.queue.maxsize))

        try:
            self.send_response(200)
            self.send_header('Content-Length', len(response))
            self.end_headers()
            self.wfile.write(response)
        except:
            print "Client disconnected before waiting for response"
            print_exception(*sys.exc_info())
            print "continuing..."

if __name__ == "__main__":
    BugsnagAgent().start()


bugsnag_agent/__init__.py
import sys

if sys.version_info[0] == 3:
    from http.server import BaseHTTPRequestHandler, HTTPServer
    from configparser import RawConfigParser
    from queue import Queue
    from _thread import interrupt_main
    from urllib.request import Request, urlopen
    from urllib.error import URLError
else:
    from BaseHTTPServer import BaseHTTPRequestHandler, HTTPServer
    from ConfigParser import RawConfigParser
    from Queue import Queue
    from thread import interrupt_main
    from urllib2 import Request, urlopen, URLError

from argparse import ArgumentParser
from threading import Thread
from time import sleep
from traceback import print_exception
import logging


logger = logging.getLogger(__name__)
log_fmt = '%(asctime)s [%(levelname)s]: %(message)s (%(filename)s: %(lineno)d)'
logging.basicConfig(format=log_fmt)


class BugsnagAgent(object):
    """
    The BugsnagAgent sits on your server and forwards exception payloads to
    https://notify.bugsnag.com/.

    It's designed to protect you against any latency spikes that may
    occur talking across the internet in an exception handler.
    """
    DEFAULTS = {
        'endpoint': 'https://notify.bugsnag.com/',
        'listen': '127.0.0.1',
        'log_level': 'INFO',
        'port': 3829,
        'queue_length': 1000
    }

    FORWARDED_HEADERS = [
        'bugsnag-sent-at',
        'bugsnag-api-key',
        'bugsnag-payload-version'
    ]

    def __init__(self):
        self.parse_config()
        self.queue = Queue(self.DEFAULTS['queue_length'])

        if self.log_level:
            logger.setLevel(getattr(logging, self.log_level.upper()))
        else:
            logger.setLevel(getattr(logging, self.DEFAULTS['log_level']))

    def parse_config(self):
        """
        Initializes self.port, self.listen, self.endpoint and self.connection
        """

        parser = ArgumentParser(description='Bugsnag Agent')
        parser.add_argument(
            '-c', '--config',
            dest='config_file',
            default='/etc/bugsnag.conf',
            help='The path to your config file (default /etc/bugsnag.conf)'
        )
        parser.add_argument(
            '-e', '--endpoint',
            dest='endpoint',
            help=("The URL of your Bugsnag server (default {})".format(
                self.DEFAULTS['endpoint']))
        )
        parser.add_argument(
            '-p', '--port',
            dest='port',
            type=int,
            help=("The port to bind to (default {})".format(
                self.DEFAULTS['port']))
        )
        parser.add_argument(
            '-i', '--ip',
            dest='listen',
            help=("The IP to listen on (use 0.0.0.0 to allow anyone to "
                  "connect , default {})".format(self.DEFAULTS['listen']))
        )
        parser.add_argument(
            '-l', '--log-level',
            dest='log_level',
            help=("Logging verbosity, default {}".format(
                self.DEFAULTS['log_level']))
        )

        args = parser.parse_args()

        config = RawConfigParser()
        config.read(args.config_file)

        # Iterate over arguments and set values in oder of precedence:
        # 1 - Arguments
        # 2 - Config file
        # 3 - Internal defaults
        conf_opts = {"port": config.getint,
                     "endpoint": config.get,
                     "listen": config.get,
                     "log_level": config.get,
                     "ip": config.get}
        for opt, _ in vars(args).items():
            if getattr(args, opt) is not None:
                setattr(self, opt, getattr(args, opt))
            elif config.has_option('bugsnag', opt) and opt in conf_opts:
                setattr(self, opt, conf_opts[opt]('bugsnag', opt))
            else:
                setattr(self, opt, self.DEFAULTS[opt])

    def start(self):
        """
        Run the agent, and wait for a SIGINT or SIGTERM.
        """
        try:
            server = Thread(target=self._thread(self._server), name='server')
            server.setDaemon(True)
            server.start()

            for _ in range(0, 10):
                client = Thread(
                    target=self._thread(self._client),
                    name='client'
                )
                client.setDaemon(True)
                client.start()

            logger.info("Bugsnag Agent started. http://{ip}:{port} -> "
                        "{endpoint}".format(
                            ip=self.listen,
                            port=self.port,
                            endpoint=self.endpoint
                        ))
            while True:
                sleep(1000)
        except KeyboardInterrupt:
            # give threads time to print exceptions
            sleep(0.1)

    def enqueue(self, body, headers={}):
        """
        Add a new payload to the queue.
        """
        try:
            self.queue.put_nowait({
                'body':body,
                'headers':headers
            })

            logger.info("Enqueued {body_length} bytes "
                        "({queue_size}/{queue_max_size})".format(
                            body_length=len(body),
                            queue_size=self.queue.qsize(),
                            queue_max_size=self.queue.maxsize
                        )
                        )
        except:
            logger.info(
                "Discarding report as queue is full: {}".format(repr(body)))

    def _server(self):
        """
        Run the HTTP server on (self.listen, self.port) that puts
        payloads into the queue.
        """
        server = HTTPServer((self.listen, self.port),
                            BugsnagHTTPRequestHandler)
        server.bugsnag = self
        server.serve_forever()

    def _client(self):
        """
        Continually monitor the queue and send anything in it to Bugsnag.
        """
        while True:
            request = self.queue.get(True)
            body = request['body']
            headers = request['headers']
            logger.info("Sending {body_length} bytes ({queue_size}/"
                "{queue_max_size})".format(
                    body_length=len(body),
                    queue_size=self.queue.qsize(),
                    queue_max_size=self.queue.maxsize
                )
            )

            try:
                req = Request(self.endpoint, body, headers)
                res = urlopen(req)
                res.read()
            except URLError as e:
                if hasattr(e, 'code') and e.code in (400, 500):
                    logger.warning('Bad response, removing report ({code}: {msg})'.format(
                        code=e.code,
                        msg=e.msg
                    ))
                else:
                    logger.warning('Cannot send request. Retrying in 5 seconds')
                    if logger.isEnabledFor(logging.DEBUG):
                        print_exception(*sys.exc_info())
                    sleep(5)
                    self.enqueue(body=body, headers=headers)

    def _thread(self, target):
        def run():
            try:
                target()
            except:
                interrupt_main()
                print_exception(*sys.exc_info())
            pass
        return run


class BugsnagHTTPRequestHandler(BaseHTTPRequestHandler):

    def do_OPTIONS(self):
        """
        Enable CORS while running on a different host
        """
        self.send_response(200, "ok")
        self.send_header('Access-Control-Allow-Origin', '*')
        self.send_header('Access-Control-Allow-Methods', 'GET, OPTIONS, POST, HEAD')
        self.send_header("Access-Control-Allow-Headers", "*")
        self.end_headers()

    def do_GET(self):
        """
        Show the current status of the agent
        """
        self.send_response(200)
        self.send_header('Access-Control-Allow-Origin', '*')
        self.end_headers()
        bugsnag = self.server.bugsnag
        self.wfile.write("Bugsnag agent: {listen}:{port} -> {endpoint} " \
            "({queue_size}/{queue_max_size})\n".format(
                listen=bugsnag.listen,
                port=bugsnag.port,
                endpoint=bugsnag.endpoint,
                queue_size=bugsnag.queue.qsize(),
                queue_max_size=bugsnag.queue.maxsize
            )
        )

    def do_POST(self):
        """
        Accept a payload for forwarding
        """
        bugsnag = self.server.bugsnag
        body = self.rfile.read(int(self.headers['Content-Length']))
        headers = {}
        for key, value in self.headers.items():
            if key.lower() in BugsnagAgent.FORWARDED_HEADERS:
                headers[key] = value

        bugsnag.enqueue(body=body, headers=headers)

        response = "OK {ip}:{port} -> {endpoint} " \
            "({queue_size}/{queue_max_size})\n".format(
                ip=bugsnag.listen,
                port=bugsnag.port,
                endpoint=bugsnag.endpoint,
                queue_size=bugsnag.queue.qsize(),
                queue_max_size=bugsnag.queue.maxsize
            )

        try:
            self.send_response(200)
            self.send_header('Access-Control-Allow-Origin', '*')
            self.send_header('Content-Length', len(response))
            self.end_headers()
            self.wfile.write(response.encode())
        except:
            logger.info('Client disconnected before waiting for response')
            print_exception(*sys.exc_info())
            logger.info('Continuing...')

if __name__ == '__main__':
    BugsnagAgent().start()


def main():
    BugsnagAgent().start()
